vllm/spec_decode/spec_decode_worker.py

```diff
@@ -107,6 +108,10 @@ def create_worker(
         else:
             proposer_worker = MultiStepWorker(**draft_worker_kwargs)
 
+        proposer_worker = SmallerTpProposerWorker.maybe_wrap_worker(
```

Comment:
What do you think of the idea of merging the logic of SmallerTpProposerWorker into MultiStepWorker?

MultiStepWorker AFAICS is always wrapped with SmallerTpProposerWorker.

Reply:
Yes, they work tightly together, and I've also considered that way.
But there're other types of draft workers such as `NGramWorker`.
So I've implemented the code to also support `NGramworker` to run with smaller tp degree.

If your goal is to only support MultiStepWorker, then it seems better to combine them.

Reply:
> Yes, they work tightly together, and I've also considered that way. But there're other types of draft workers such as `NGramWorker`. So I've implemented the code to also support `NGramworker` to run with smaller tp degree.
> 
> If our (vllm's) goal is to only support MultiStepWorker, then it seems better to combine them.

My understanding is that `NGramworker` is not [using tensor parallelism, ](https://github.com/vllm-project/vllm/blob/0f0d8bc065f3608e7657a9696f5d2d7c0d6722d1/vllm/spec_decode/ngram_worker.py#L37)since it's only doing prompt token lookup.

There are two new speculative workers being worked on:
https://github.com/vllm-project/vllm/pull/4978 (Medusa), https://github.com/vllm-project/vllm/pull/4947 (MLPSpeculator).

In #4947, `MLPSpeculatorWorker` reuses `MultiStepWorker`. On the Medusa side, since it doesn't run the draft model separately, there is no concept of  `speculative_tensor_parallel_size`.

So I think it's cleaner to put the logic in `MultiStepWorker`.

Reply:
Oh, I didn't know `NGramWorker` does not use tp..!
Then I completely agree with your suggestion.

Thank you for pointing out the important part! 👍 
I'll let you know after refactoring.

Reply:
Just FYI for #4947 (MLPSpeculator) we haven't fixed the TP support yet but that's the last thing to do (working on it now). We suspect it won't make sense to run the speculator model with TP so priority will be fixing target model TP but we may wire up TP for the speculator too so that we can at least test it out.